cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java

```diff
@@ -17,49 +17,122 @@
+import co.cask.tephra.Transaction;
+import com.google.common.base.Charsets;
 import com.google.common.base.Preconditions;
-import com.google.common.base.Predicate;
+import com.google.common.collect.HashBasedTable;
 import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.ImmutableSortedMap;
-import com.google.common.collect.Maps;
+import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.primitives.Ints;
 import com.google.common.primitives.Longs;
+import com.google.gson.Gson;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
 import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import javax.annotation.Nullable;
 
 /**
  * Class for persisting consumer state per queue consumer.
  */
 final class HBaseConsumerStateStore extends AbstractDataset implements QueueConfigurer {
 
+  private static final Gson GSON = new Gson();
+
   private final QueueName queueName;
   private final Table table;
-  private final byte[] minStartRow;
+  private final byte[] barrierScanStartRow;
+  private final byte[] barrierScanEndRow;
+  private Transaction transaction;
 
   HBaseConsumerStateStore(String datasetName, QueueName queueName, Table table) {
     super(datasetName, table);
     this.queueName = queueName;
     this.table = table;
-    this.minStartRow = QueueEntryRow.getQueueEntryRowKey(queueName, 0L, 0);
+    this.barrierScanStartRow = Bytes.add(queueName.toBytes(), QueueEntryRow.getQueueEntryRowKey(queueName, 0L, 0));
+    this.barrierScanEndRow = Bytes.stopKeyForPrefix(
+      Bytes.add(queueName.toBytes(), QueueEntryRow.getQueueEntryRowKey(queueName, Long.MAX_VALUE, 0)));
+  }
+
+  @Override
+  public void startTx(Transaction tx) {
+    super.startTx(tx);
+    this.transaction = tx;
   }
 
   /**
    * Returns the consumer state as stored in the state store for the given consumer.
    */
   HBaseConsumerState getState(long groupId, int instanceId) {
-    byte[] stateColumn = getConsumerStateColumn(groupId, instanceId);
-    byte[] startRow = table.get(queueName.toBytes(), stateColumn);
-    Preconditions.checkState(startRow != null,
-                             "No consumer state found. GroupId: %s, InstanceId: %s", groupId, instanceId);
+    ConsumerState consumerState = getConsumerState(groupId, instanceId);
+    QueueBarrier startBarrier = consumerState.getStartBarrier();
+    QueueBarrier endBarrier = consumerState.getEndBarrier();
+    if (startBarrier == null && endBarrier == null) {
+      throw new IllegalStateException(
+        String.format("Unable to find barrier information for consumer. Queue: %s, GroupId: %d, InstanceId:%d",
+                      queueName, groupId, instanceId));
+    }
+
+    ConsumerGroupConfig groupConfig = startBarrier != null ? startBarrier.getGroupConfig()
```

Comment:
I am trying to understand... Is it correct that:
- all consumers now have a state with a startBarrier (starting at 0) 
- so if the startBarrier is null, then we are in compatibility mode, consuming entries that were produced with the old strategy
- every time the config changes, a new barrier is added. That means if the state has an end barrier, then the consumer must wait for all other consumers to catch up when he reaches the barrier.

Sometimes it is useful to explain that in Javadocs or comments. The rest of the code gets much easier to understand